source/extensions/filters/network/thrift_proxy/conn_manager.cc

```diff
@@ -537,23 +544,22 @@ ConnectionManager::ActiveRpc::applyFilters(FilterType* filter,
   }
 
   filter_action_ = nullptr;
-  filter_context_.reset();
+  filter_context_ = absl::monostate();
 
   return FilterStatus::Continue;
 }
 
-void ConnectionManager::ActiveRpc::prepareFilterAction(DecoderEvent event, absl::any data) {
+void ConnectionManager::ActiveRpc::prepareFilterAction(DecoderEvent event) {
   // DecoderEvent::ContinueDecode indicates we're handling previous filter action with the
-  // filter chain. Therefore, we should not reset filter_action_ and filter_context_.
+  // filter chain. Therefore, we should not reset filter_action_.
   if (event == DecoderEvent::ContinueDecode) {
     return;
   }
-  filter_context_ = data;
 
   switch (event) {
   case DecoderEvent::TransportBegin:
     filter_action_ = [this](DecoderEventHandler* filter) -> FilterStatus {
-      MessageMetadataSharedPtr metadata = absl::any_cast<MessageMetadataSharedPtr>(filter_context_);
+      MessageMetadataSharedPtr metadata = absl::get<MessageMetadataSharedPtr>(filter_context_);
```

Comment:
It seems to me that the only reason that values are being stored in filter_context_ is to pass them into these lambdas. So it seems like this function could accept a parameter `FilterContext& data` (instead of `absl::any data`) and you can trigger the copy that's necessary for the string values and shared pointers in this method via `filter_context_ = data`.

Further, I think we might even be able to put the copy in the lambda capture and get rid of `filter_context_` altogether. Given a `FilterContext &data` parameter, I think defining the lambdas as
```
[copy = FilterContext(data)](DecoderEventHandler* filter) -> FilterStatus { ... }
```
should do it. It might also be possible to use `FilterContext&& data` and `std::move(data)`.

Reply:
On the other hand, thrift filter should be able to modify the data, so modification the capture list would cause
`error: increment of read-only variable `, which is another reason to keep `filter_context_` to know the current FilterContext value during the filter chain.

Reply:
Current version should fit your great review feedback. Thanks for your great input and please take another look :)